### DISCUSSION TITLE
Chained resources vs. Suspense

### DISCUSSION THREAD

**Discussion by matburton** [solidjs/solid#1015] [Q&A] [answered]
I wrote a component which performs two fetches, the second depending on the first.
Hence as a n00b I wrote code like below.

```TypeScript
import { createResource, Suspense } from "solid-js";

export default () => {

    const delay = (ms: number) => new Promise(r => setTimeout(r, ms));

    const [resourceA] = createResource(async () => {

        await delay(2000); // Pretend this is a slow fetch

        return "a";
    });

    const [resourceB] = createResource(resourceA, async a => {

        await delay(2000);

        return `${a}-b`;
    });

    return <Suspense fallback={<em>Loading</em>}><p>{resourceB()}</p></Suspense>;
};
```

This works as intended in terms of the ordering and end result.

However `Loading` is not shown until the fetch begins for `resourceA`.
So for the frist two seconds nothing is rendered.

Is this intended? Should I have expressed this in another way?

Should I be explicitly indicating the dependency with `on` somewhere?
(I tried but couldn't work this out)

The only thing I found that worked was instead writing:
```TypeScript
const getResourceB = () => { resourceA(); return resourceB(); };

return <Suspense fallback={<em>Loading</em>}><p>{getResourceB()}</p></Suspense>;
```

But that feels too messy to be the correct approach?
(Maybe it's the only way to work with both Suspense and ErrorBoundary?)

**Comment by lxsmnsyc** [comment, marked as answer]
> Is this intended?

Yes, because the suspending mechanism  happens on resource reads. Take note that since only `resourceB` is read inside the Suspense, `resourceA` would suspend outside of the Suspense.

> Should I have expressed this in another way?

In which case you need to wrap your entire component with Suspense.
